### PR TITLE
Move platform specific information to `PlatformConfiguration` class

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This moves most of the logic that is currently in `Window` into a new `PlatformConfiguration` class, in preparation for multi-window support.

I split the platform specific configuration out of the `Window` class, into a separate `PlatformConfiguration` class that keeps the configuration of the platform, and manages a lot of the platform-specific communication that `Window` used to manage (things like `DispatchPlatformMessage` and `UpdateLocales`, etc.), since these things don't really belong in a window, they belong to the platform, and when you have more than one window, keeping them on each window doesn't make sense. The `PlatformConfiguration` class will connect to the `PlatformDispatcher` on the Dart side once that class exists.

Most of the new `PlatformConfiguration` class consists of methods moved verbatim from `Window`. I also renamed `WindowData` to `PlatformData`, since it no longer applies to windows. In the future, I'll be migrating the framework to match, so that there is a `PlatformDispatcher` class that takes care of communicating with `PlatformConfiguration`. [I didn't call them the same thing, because the engine side is meant to be the source of truth, and the framework side (once it exists) is meant to dispatch messages to and from this source of truth.]

## Related Issues

- flutter/flutter#30701

## Tests

- Added tests for new platform configuration class.

### CODE REVIEW

File: shell/testing/tester_main.cc [hunk 1]
```diff
@@ -234,11 +234,8 @@ int RunTester(const flutter::Settings& settings,
                      }
                    });
 
-  flutter::ViewportMetrics metrics;
-  metrics.device_pixel_ratio = 3.0;
-  metrics.physical_width = 2400;   // 800 at 3x resolution
-  metrics.physical_height = 1800;  // 600 at 3x resolution
-  shell->GetPlatformView()->SetViewportMetrics(metrics);
+  // 800x600 at 3x resolution.
+  shell->GetPlatformView()->SetViewportMetrics({3.0, 2400, 1800});
```
Comment (stuartmorgan) [inline shell/testing/tester_main.cc:238]: Anonymous struct initialization is fragile to struct changes; please use the explicit structure used before (although it should initialize the struct to `{}`, which was missing).

File: shell/platform/embedder/embedder.h [hunk 2]
```diff
@@ -1010,6 +1061,9 @@ FlutterEngineResult FlutterEngineCreateAOTData(
 FLUTTER_EXPORT
 FlutterEngineResult FlutterEngineCollectAOTData(FlutterEngineAOTData data);
 
+/// The order, type, and size of these struct members must remain the same, and
+/// members should not be removed. This is to allow for forward and backward
+/// compatibility between the engine and the embedder.
```
Comment (stuartmorgan) [inline shell/platform/embedder/embedder.h:1066]: It seems odd, and noisy, to put this boilerplate everywhere. It's also odd to explicitly call this out N times, when don't mention that function names can't change, argument numbers/orders/types can't change, core behavior of functions can't change, etc. I'm not sure why this particular piece of ABI preservation needs to be called out specially.

A file-level comment that says that this header presents a stable ABI, and gives examples like function and struct changes, seems like a better option if we want to comment thing like this.

@chinmaygarde for thoughts.

File: lib/ui/window/window.h [hunk 3]
```diff
@@ -9,102 +9,26 @@
...
 
-  void DidCreateIsolate();
-  void UpdateWindowMetrics(const ViewportMetrics& metrics);
-  void UpdateLocales(const std::vector<std::string>& locales);
-  void UpdateUserSettingsData(const std::string& data);
-  void UpdateLifecycleState(const std::string& data);
-  void UpdateSemanticsEnabled(bool enabled);
-  void UpdateAccessibilityFeatures(int32_t flags);
-  void DispatchPlatformMessage(fml::RefPtr<PlatformMessage> message);
-  void DispatchPointerDataPacket(const PointerDataPacket& packet);
```
Comment (stuartmorgan) [inline lib/ui/window/window.h:86]: Isn't pointer-data per-window?

File: lib/ui/window/platform_configuration.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+enum class AccessibilityFeatureFlag : int32_t {
+  kAccessibleNavigation = 1 << 0,
+  kInvertColors = 1 << 1,
+  kDisableAnimations = 1 << 2,
+  kBoldText = 1 << 3,
+  kReduceMotion = 1 << 4,
+  kHighContrast = 1 << 5,
+};
+
+class PlatformConfigurationClient {
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration.h:51]: This should have a class-level comment describing what it is/does.

File: lib/ui/window/platform_configuration.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+ public:
+  virtual std::string DefaultRouteName() = 0;
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration.h:53]: These should really all have comments. It's a pre-existing issue though, and guess is should be cleaned up in a docs sweep at some point.

File: lib/ui/window/platform_configuration.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  virtual void HandlePlatformMessage(fml::RefPtr<PlatformMessage> message) = 0;
+  virtual FontCollection& GetFontCollection() = 0;
+  virtual void UpdateIsolateDescription(const std::string isolate_name,
+                                        int64_t isolate_port) = 0;
+  virtual void SetNeedsReportTimings(bool value) = 0;
+  virtual std::shared_ptr<const fml::Mapping> GetPersistentIsolateData() = 0;
+  virtual std::unique_ptr<std::vector<std::string>>
+  ComputePlatformResolvedLocale(
+      const std::vector<std::string>& supported_locale_data) = 0;
+
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration.h:66]: This class should probably be preventing copies (deleting the relevant constructors).

File: lib/ui/window/platform_configuration.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+ protected:
+  virtual ~PlatformConfigurationClient();
+};
+
+class PlatformConfiguration final {
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration.h:71]: Needs a class-level comment as well.

File: lib/ui/window/platform_configuration.h [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  ///
+  /// @return the Window.
+  const Window& window() const { return *window_; }
+
+  //----------------------------------------------------------------------------
+  /// @brief Sets the viewport metrics of the Window.
+  ///
+  /// @param[in] window_metrics The viewport metrics to replace the old ones
+  ///            with.
+  void SetWindowMetrics(const ViewportMetrics& window_metrics);
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration.h:110]: Shouldn't this just be a Window method? I'm not sure why we need the passthrough.

File: lib/ui/window/platform_configuration.h (hunk 4 above)
Comment (gspencergoog) [reply lib/ui/window/platform_configuration.h:53]: Yeah, I agree. I was just moving it, and it didn't have any comments before, but you're right.

File: lib/ui/window/platform_configuration.h (hunk 6 above)
Comment (gspencergoog) [reply lib/ui/window/platform_configuration.h:68]: Done.

File: lib/ui/window/platform_configuration.h (hunk 5 above)
Comment (gspencergoog) [reply lib/ui/window/platform_configuration.h:55]: Done.

File: lib/ui/window/platform_configuration.h (hunk 4 above)
Comment (gspencergoog) [reply lib/ui/window/platform_configuration.h:53]: I put some comments in, but they need formatting to whatever the correct form is for public docs (doxygen?)

Comment (gspencergoog) [reply lib/ui/window/platform_configuration.h:62]: OK, they're in and formatted for doxygen.

File: lib/ui/window/platform_configuration_unittests.cc [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+namespace flutter {
+namespace testing {
+
+class DummyPlatformConfigurationClient : public PlatformConfigurationClient {
+ public:
+  DummyPlatformConfigurationClient() {
+    std::vector<uint8_t> data;
+    isolate_data_.reset(new ::fml::DataMapping(data));
+  }
+  virtual std::string DefaultRouteName() { return "TestRoute"; }
```
Comment (stuartmorgan) [inline lib/ui/window/platform_configuration_unittests.cc:28]: These should all be marked `override` rather than `virtual`.

Comment (gspencergoog) [reply lib/ui/window/platform_configuration_unittests.cc]: Good point.  Fixed.

**Comment by chinmaygarde** [comment]
Can we land this?

**Comment by gspencergoog** [comment]
@chinmaygarde Yes, I was waiting for fixes to the lint errors in another PR (and the tree was all hosed for a while).

I'll try and land this today or tomorrow.

**Comment by gspencergoog** [comment]
Sorry, still trying to land this. I had to merge conflicts.

**Pull request merged**